src/com/goide/inspections/GoStructTagInspection.java

```diff
@@ -0,0 +1,107 @@
+ * limitations under the License.
+ */
+
+package com.goide.inspections;
+
+import com.goide.psi.GoFieldDeclaration;
+import com.goide.psi.GoStructType;
+import com.goide.psi.GoTag;
+import com.goide.psi.GoVisitor;
+import com.intellij.codeInspection.LocalInspectionToolSession;
+import com.intellij.codeInspection.ProblemsHolder;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Implements <a href="https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/1983"/>, an
+ * inspector that warns if a go StructTag is not well-formed according to Go language conventions.
+ */
+public class GoStructTagInspection extends GoInspectionBase {
+  @NotNull
+  @Override
+  protected GoVisitor buildGoVisitor(@NotNull final ProblemsHolder holder,
+                                     @SuppressWarnings({"UnusedParameters", "For future"}) @NotNull LocalInspectionToolSession session) {
+    return new GoVisitor() {
+      @Override
+      public void visitStructType(@NotNull GoStructType o) {
+        for (GoFieldDeclaration field : o.getFieldDeclarationList()) {
+          if (field.getTag() == null) {
+            continue;
+          }
+          if (!isValidTag(field.getTag())) {
+            holder.registerProblem(field.getTag(), "Bad syntax for struct tag value");
+          }
+        }
+      }
+    };
+  }
+
+  // Implementation based on validateStructTag from the go vet tool:
+  // https://github.com/golang/tools/blob/master/cmd/vet/structtag.go.
+  private boolean isValidTag(GoTag tag) {
+    StringBuilder tagText = new StringBuilder(tag.getText().substring(1, tag.getTextLength() - 1));
+    while (tagText.length() > 0) {
+      int i;
+      i = 0;
+
+      // Skip leading spaces.
+      while (i < tagText.length() && tagText.charAt(i) == ' ') {
+        i++;
+      }
+
+      tagText.delete(0, i);
+      if (tagText.length() == 0) {
+        return true;
+      }
+
+      // Scan to colon. A space, a quote or a control character is a syntax error.
+      // Strictly speaking, control chars include the range [0x7f, 0x9f], not just
+      // [0x00, 0x1f], but in practice, we ignore the multi-byte control characters
+      // as it is simpler to inspect the tag's bytes than the tag's runes.
+      i = 0;
+      while (i < tagText.length() &&
+             tagText.charAt(i) > ' ' &&
+             tagText.charAt(i) != ':' &&
+             tagText.charAt(i) != '"' &&
+             tagText.charAt(i) != 0x7f) {
+        i++;
+      }
+      if (i == 0 || (i + 1) > tagText.length() || tagText.charAt(i) != ':') {
+        return false;
+      }
+      tagText.delete(0, i + 1);
+
+      // Scan quoted string to find value.
+      i = 1;
+      while (i < tagText.length() && tagText.charAt(i) != '"') {
+        if (tagText.charAt(i) == '\\') {
+          i++;
+        }
+        i++;
+      }
+
+      if (i >= tagText.length()) {
+        return false;
+      }
+
+      // TODO(sjamesr): check strconv.Unquote(quotedValue)
```

Comment:
Do you think it's good to submit as-is with the TODO in place? I think in practice the current check will do quite a good job.